### ISSUE TITLE
Build out and document contributing rules and QA process

### ISSUE THREAD

**Issue by alexkyllo** [ScienceCommons/api#163]
1. Deploy a second instance of the api app to Heroku to use as a staging and User Acceptance Testing environment - *In Progress*
2. Create a CONTRIBUTING.md document to list rules and steps for contributing code to this repository
3. Build out a unit test suite for the business logic in the client-side Javascript code
4. Document a series of manual UI regression tests to be run before deploying to production
5. Automate the UI tests using a Selenium-based browser automation framework
6. Set up Travis CI to automatically detect new commits to the repo, check out the code, run tests, and deploy.

**Comment by alexkyllo** [comment]
Staging site is up and running at https://curatescience-staging.herokuapp.com/beta

**Comment by alexkyllo** [comment]
@eplebel since you have admin rights, would you please go to https://travis-ci.org/profile/ScienceCommons (log in with your Github account) and click the button to enable ScienceCommons/www for Travis CI?

After you do that, I can set up Travis CI to watch the github repo for commits and deploy to staging automatically.

I will create a "staging" branch in each github repo and we can require developers to submit pull requests to that branch. Then all you'll have to do is merge the pull request and Travis will automatically grab the code, build the app, run any tests, and deploy to https://curatescience-staging.herokuapp.com/beta so you can UAT it.

![capture](https://cloud.githubusercontent.com/assets/2609312/7567564/a8ab2104-f7b2-11e4-961f-9dfb8548bd2e.JPG)

**Comment by eplebel** [comment]
Great! I just authorized Travis CI to have access to our github repos, let me know if you need anything else!

**Comment by alexkyllo** [comment]
Perfect, I can see both repos in the Travis interface now, thanks! I'll work on creating the build/test/deploy scripts.

**Comment by alexkyllo** [comment]
The api repo will now deploy to heroku staging when a pull request is merged into the staging branch, and it will deploy to heroku production when a pull request is merged into the master branch.

Working on setting up a similar process for the www repo.

**Comment by eplebel** [comment]
Sweet, that's awesome!! 

So this effectively solves the Elance deploy problem we discussed on Sunday right (given I could merge PRs from an Elancer into the staging branch)?

**Comment by alexkyllo** [comment]
Yes, exactly--but currently only for the api repo, I'm still working on setting it up for www.

The Elancer will just need to be careful to open their pull requests on the staging branches, and you'll also want to be careful not to merge a pull request into master until you've merged it into staging and tested it first.

I'll let you know as soon as everything is set up and ready to fully switch over to this workflow. And I'll deploy Tetiana's latest pull requests to staging manually tonight.

**Comment by eplebel** [comment]
Great, that's so awesome, i'm so excited!!
But you're right, I'll definitely need to be careful.

And great re: Tetiana's latest PRs; looks like there are a few conflicts
with the api PR, probably having to do with the new staging stuff you
added, but I suspect this should be fairly straightforward to address.

Thanks a million
![image: Image result for thanks a million]

**Comment by alexkyllo** [comment]
TravisCI now successfully deploys the `www` repo code to staging environment when it detects a commit to `staging` branch.
